### PR TITLE
Dashboards: Remove slug leftovers - replace with `/d/:uid` endpoint

### LINKED ISSUE #35447: Dashboards: Dashboards in Application's `plugin.json` stopped working after upgrading to v8
**What happened**:

As per the discussion on the public Grafana Slack ([link](https://grafana.slack.com/archives/C3HJV5PNE/p1623251539076500)), app plugins stopped working in Grafana 8.0.0.

This is a consequence of https://github.com/grafana/grafana/pull/35104 which removed support of endpoints that contained slug, but a leftover remained:

https://github.com/grafana/grafana/blob/e3afb63e62d5fb7ec577c33efaa7473de37aff45/pkg/plugins/app_plugin.go#L119
and

https://github.com/grafana/grafana/blob/e3afb63e62d5fb7ec577c33efaa7473de37aff45/pkg/api/index.go#L111

**What you expected to happen**:
Application dashboards to work properly in Grafana v8.0.0.

**How to reproduce it (as minimally and precisely as possible)**:
Comment on the public Slack says:

> the dashboard is loaded and works, but they are not accessible using the menu - something was changed or should be updated

which is expected since it can't find the correct slug.

**Environment**:
- Grafana version: v8.0.0

### PR DESCRIPTION
**Pull request by dsotirakis**
**What this PR does / why we need it**:

This PR is raised as a potential solution for #35447 . After `slug` removal, needed to make modifications in remaining leftovers.

**Which issue(s) this PR fixes**:

Fixes #35447

**Special notes for your reviewer**:

Replicated the issue using redis-app plugin:

(_check at the ref links at the bottom left_)

* Current behaviour (`main branch`):

Untitled

* Behaviour after fix (`current PR`):

Kapture 2021-06-10 at 15 07 24

Example when clicking to one of those links:

Kapture 2021-06-10 at 15 09 52

### CODE REVIEW

File: pkg/api/index.go [hunk 1]
```diff
@@ -109,8 +112,22 @@ func (hs *HTTPServer) getAppLinks(c *models.ReqContext) ([]*dtos.NavLink, error)
...
+					return nil, err
+				}
+				var dashboard models.Dashboard
+				jsonParser := json.NewDecoder(jsonBody)
+				if err = jsonParser.Decode(&dashboard); err != nil {
+					return nil, err
+				}
 				link := &dtos.NavLink{
-					Url:  hs.Cfg.AppSubURL + "/dashboard/db/" + include.Slug,
+					Url:  hs.Cfg.AppSubURL + "/d/" + dashboard.Uid,
```
Comment (marefr) [inline pkg/api/index.go:130]: It doesn't feel optimal to read the dashboard.json files from disk each time requesting the index page and in addition having this kind of code in multiple places creates code duplication. 

If you could run this code when the app is installed instead you could potentially populate a new field `include.UID`. 

However, I see one potential problem with this approach and that is that it's not guarantees that a plugin dashboard would actually include a `uid` which means that Grafana generates the `uid` when the app dashboards are saved to the database.

**Review by marefr — COMMENTED** [review]
Added a comment for now. Don't really have any perfect solution to this currently so lets continue discussions tomorrow

File: pkg/api/index.go (hunk 1 above)
Comment (dsotirakis) [reply pkg/api/index.go:130]: > If you could run this code when the app is installed instead you could potentially populate a new field include.UID.

Agree with this, but this means that we'll have to make changes to our `plugin.json` files, right?

> However, I see one potential problem with this approach and that is that it's not guarantees that a plugin dashboard would actually include a uid which means that Grafana generates the uid when the app dashboards are saved to the database.

Hm, then that means that the actual endpoints `dashboards/uid/:uid` may not work at times as well? 🤔

Comment (dsotirakis) [reply pkg/api/index.go:130]: @torkelo do you happen to have any thoughts on that?

**Pull request closed without merging**